pmd-ui/src/main/java/net/sourceforge/pmd/util/fxdesigner/AuxClassPathController.java

```diff
@@ -0,0 +1,111 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+
+package net.sourceforge.pmd.util.fxdesigner;
+
+import java.io.File;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ResourceBundle;
+
+import net.sourceforge.pmd.util.fxdesigner.util.DesignerUtil;
+import net.sourceforge.pmd.util.fxdesigner.util.beans.SettingsOwner;
+
+import javafx.fxml.FXML;
+import javafx.fxml.FXMLLoader;
+import javafx.fxml.Initializable;
+import javafx.scene.Parent;
+import javafx.scene.Scene;
+import javafx.scene.control.Button;
+import javafx.scene.control.ListView;
+import javafx.stage.FileChooser;
+import javafx.stage.Modality;
+import javafx.stage.Stage;
+
+public class AuxClassPathController implements Initializable, SettingsOwner {
+
+    private final DesignerRoot designerRoot;
+    private List<File> list = new ArrayList<>();
+    @FXML
+    private Button removeFiles;
+    @FXML
+    private Button selectFile;
+    @FXML
+    private ListView<File> fileList;
+    @FXML
+    private final MainDesignerController parent;
+
+
+    public AuxClassPathController(DesignerRoot designerRoot, MainDesignerController parent) {
+        this.designerRoot = designerRoot;
+        this.parent = parent;
+    }
+
+
+    @Override
+    public void initialize(URL location, ResourceBundle resources) {
+
+        selectFile.setOnAction(e -> fileSelected());
+        removeFiles.setOnAction(e -> setRemoveFiles());
+
+    }
+
+
+    private void fileSelected() {
+        FileChooser chooser = new FileChooser();
+        chooser.setTitle("Select Files");
+        chooser.getExtensionFilters().addAll(
+            new FileChooser.ExtensionFilter("Java JARs", "*.jar"),
+            new FileChooser.ExtensionFilter("Java WARs", "*.war"),
+            new FileChooser.ExtensionFilter("Java EARs", "*.ear"),
+            new FileChooser.ExtensionFilter("Java class files", "*.class")
+        );
+        File file = chooser.showOpenDialog((designerRoot.getMainStage()));
+        list.add(file);
+        displayFiles();
+    }
+
+
+    private void displayFiles() {
+
+        for (int i = 0; i < list.size(); i++) {
+            fileList.refresh();
+            fileList.getItems().add(new File(String.valueOf(list.get(i))));
```

Comment:
Lol wtf

You're basically readding all the items that are already there. Did you try your code?

You don't need to do any of that. Remove this `displayFiles` method

Reply:
Yes I did and yeah I am trying to figure out to stop those multiple entry display.

Reply:
So, each time you select a file, you add it to the list (line 67) and call this method. Here you go over the whole list and add each entry back to `fileList`. In other words:

1. first time, `list` is empty, single entry is added, you come here, and get 1 item in the list view
2. second time, `list` has the previous element, and a second one is aded. You come here, and add again items 1 and 2 (1 gets repeated!)

If you follow Clément's advice, and drop the `list`, you can remove this method completely. Just add the file to `fileList.getItems()` in line 67 (instead of the separate list) and at most you will have to add a call to `fileList.refresh();` (not sure it's actually needed, but easy to test).

Reply:
Yes the refresh call is not needed. The ListView (and other controls like TreeView and TableView) stores its items in an [ObservableList](https://docs.oracle.com/javase/8/javafx/api/javafx/collections/ObservableList.html?is-external=true), which means adding, removing, and reordering elements trigger events which you can subscribe to. The ListView subscribes to these and refreshes its UI whenever a change is notified.